python/cufinufft/tests/test_basic.py

```diff
@@ -7,21 +9,30 @@
 
 import utils
 
+# NOTE: Tests below fail for shapes of size (16, 16) and tolerances 1e-4.
 
-def _test_type1(dtype, shape=(16, 16, 16), M=4096, tol=1e-3, output_arg=True):
-    complex_dtype = utils._complex_dtype(dtype)
+DTYPES = [np.float32, np.float64]
+SHAPES = [(64,), (64, 64), (64, 64, 64)]
+MS = [256, 1024, 4096]
+TOLS = [1e-2, 1e-3]
+OUTPUT_ARGS = [False, True]
 
-    dim = len(shape)
+@pytest.mark.parametrize("dtype", DTYPES)
+@pytest.mark.parametrize("shape", SHAPES)
+@pytest.mark.parametrize("M", MS)
+@pytest.mark.parametrize("tol", TOLS)
+@pytest.mark.parametrize("output_arg", OUTPUT_ARGS)
+def test_type1(dtype, shape, M, tol, output_arg):
+    complex_dtype = utils._complex_dtype(dtype)
 
-    k = utils.gen_nu_pts(M, dim=dim).astype(dtype)
-    c = utils.gen_nonuniform_data(M).astype(complex_dtype)
+    k, c = utils.type1_problem(dtype, shape, M)
 
     k_gpu = gpuarray.to_gpu(k)
     c_gpu = gpuarray.to_gpu(c)
 
     plan = Plan(1, shape, eps=tol, dtype=complex_dtype)
 
-    plan.setpts(k_gpu[0], k_gpu[1], k_gpu[2])
+    plan.setpts(*k_gpu)
```

Comment:
has setpts changed to expect a single pointer now (I was expecting kx,ky,kz as before) ?

Reply:
So there are a few things going on here. First, the `*` notation acts here to unpack a list (see [here](https://docs.python.org/3.6/tutorial/controlflow.html#unpacking-argument-lists)). The other thing is that multidimensional arrays are treated as lists along their first dimension.

Here we have `k_gpu`, which is an array of shape `(d, M)`, where `d` is the dimension and `M` is the number of points. The above expansion will then yield
```
plan.setpts(k_gpu[0])
```
for `d = 1`,
```
plan.setpts(k_gpu[0], k_gpu[1])
```
for `d = 2`, and
```
plan.setpts(k_gpu[0], k_gpu[1], k_gpu[2])
```
for `d = 3`.